File: mobile-mvp/daemon/buddy_daemon/state.py
"""BuddyState + pure reducer. Single source of truth at runtime.

Mirrors ``../../schema/state.ts``. The reducer is the only code that
mutates state. Everything else reads.

Contract (the invariants in ../README.md):

1. ``version`` is monotonic; ``reduce`` bumps by exactly 1 on change,
   returns the *same* state object otherwise.
2. Reducer is pure: no logging, no time lookup, no I/O. ``at`` is
   always passed in.
3. Prompt replacement is atomic: a heartbeat either keeps the current
   prompt (matching id), replaces it with a new id, or clears it to
   ``None``. Decisions live on ``prompt.decidedBy`` until upstream
   confirms the prompt gone (next heartbeat without it).
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field, replace
from typing import Any, Literal

from .events import (
    BleConnected,
    BleDisconnected,
    ClientDecide,
    Event,
    Heartbeat,
    OwnerSet,
    StaleTick,
    TimeSync,
)

PetState = Literal["sleep", "idle", "busy", "attention", "celebrate", "dizzy", "heart"]
DesktopStatus = Literal["disconnected", "connected", "stale"]

STALE_MS = 30_000  # must match Config.stale_timeout_ms default


@dataclass(frozen=True, slots=True)
class DesktopLink:
    status: DesktopStatus = "disconnected"
    lastHeartbeatAt: int | None = None
    secure: bool = False


@dataclass(frozen=True, slots=True)
class SessionCounts:
    total: int = 0
    running: int = 0
    waiting: int = 0


@dataclass(frozen=True, slots=True)
class PendingDecision:
    clientId: str
    decision: Literal["once", "deny"]
    at: int


@dataclass(frozen=True, slots=True)
class Prompt:
    id: str
    tool: str
    hint: str
    arrivedAt: int
    source: str = "other"  # "cursor" | "claude-code" | "other"
    decidedBy: PendingDecision | None = None


@dataclass(frozen=True, slots=True)
class Pet:
    state: PetState = "sleep"
    oneShotUntil: int | None = None
    species: Literal["bufo"] = "bufo"


@dataclass(frozen=True, slots=True)
class BuddyState:
    version: int = 0
    updatedAt: int = 0
    desktop: DesktopLink = field(default_factory=DesktopLink)
    sessions: SessionCounts = field(default_factory=SessionCounts)
    msg: str = ""
    entries: tuple[str, ...] = ()
    tokens: int = 0
    tokensToday: int = 0
    prompt: Prompt | None = None
    pet: Pet = field(default_factory=Pet)
    owner: str = ""

    def to_dict(self) -> dict[str, Any]:
        d = asdict(self)
        d["entries"] = list(self.entries)
        return d


INITIAL_STATE = BuddyState()


# ---------------------------------------------------------------------------
# Reducer
# ---------------------------------------------------------------------------


def reduce(state: BuddyState, event: Event) -> BuddyState:
    """Pure state transition. Returns same object if nothing changed."""
    new = _reduce_inner(state, event)
    if new is state:
        return state
    return replace(
        new,
        version=state.version + 1,
        updatedAt=_event_at(event) or new.updatedAt,
    )


def _reduce_inner(state: BuddyState, event: Event) -> BuddyState:
    if isinstance(event, BleConnected):
        if state.desktop.status == "connected":
            return state
        return replace(
            state,
            desktop=DesktopLink(
                status="connected",
                lastHeartbeatAt=state.desktop.lastHeartbeatAt,
                secure=state.desktop.secure,
            ),
            pet=_derive_pet(state, state.prompt, state.sessions),
        )

    if isinstance(event, BleDisconnected):
        return replace(
            state,
            desktop=DesktopLink(
                status="disconnected",
                lastHeartbeatAt=state.desktop.lastHeartbeatAt,
                secure=False,
            ),
            pet=_derive_pet_disconnected(),
        )

    if isinstance(event, StaleTick):
        if state.desktop.lastHeartbeatAt is None:
            return state
        age = event.at - state.desktop.lastHeartbeatAt
        desired: DesktopStatus = (
            "stale" if age > STALE_MS else state.desktop.status
        )
        if desired == state.desktop.status:
            return state
        return replace(
            state,
            desktop=replace(state.desktop, status=desired),
            pet=_derive_pet_for(state.prompt, state.sessions, connected=False),
        )

    if isinstance(event, TimeSync):
        # Treat as a liveness signal but otherwise state-free for MVP.
        return replace(
            state,
            desktop=replace(
                state.desktop, status="connected", lastHeartbeatAt=event.at
            ),
        )

    if isinstance(event, OwnerSet):
        if state.owner == event.name:
            return state
        return replace(state, owner=event.name)

    if isinstance(event, Heartbeat):
        sessions = SessionCounts(
            total=event.total, running=event.running, waiting=event.waiting
        )
        prompt = _reconcile_prompt(state.prompt, event)
        return replace(
            state,
            desktop=replace(
                state.desktop,
                status="connected",
                lastHeartbeatAt=event.at,
            ),
            sessions=sessions,
            msg=event.msg,
            entries=tuple(event.entries),
            tokens=event.tokens,
            tokensToday=event.tokens_today,
            prompt=prompt,
            pet=_derive_pet_for(prompt, sessions, connected=True),
        )

    if isinstance(event, ClientDecide):
        # The bridge is responsible for validation (matching promptId,
        # race handling). The reducer records the outcome only.
        if state.prompt is None or state.prompt.id != event.prompt_id:
            return state
        if state.prompt.decidedBy is not None:
            return state
        new_prompt = replace(
            state.prompt,
            decidedBy=PendingDecision(
                clientId=event.client_id,
                decision=event.decision,
                at=event.at,
            ),
        )
        return replace(state, prompt=new_prompt)

    return state


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _reconcile_prompt(prev: Prompt | None, hb: Heartbeat) -> Prompt | None:
    if hb.prompt_id is None:
        return None
    if prev is not None and prev.id == hb.prompt_id:
        return prev
    return Prompt(
        id=hb.prompt_id,
        tool=hb.prompt_tool or "",
        hint=hb.prompt_hint or "",
        arrivedAt=hb.at,
        source=hb.prompt_source or "other",
        decidedBy=None,
    )


def _derive_pet(
    state: BuddyState, prompt: Prompt | None, sessions: SessionCounts
) -> Pet:
    return _derive_pet_for(
        prompt, sessions, connected=state.desktop.status == "connected"
    )


def _derive_pet_for(
    prompt: Prompt | None, sessions: SessionCounts, *, connected: bool
) -> Pet:
    # Mirrors derive() in src/main.cpp. Celebrate/dizzy/heart are
    # one-shots driven externally and are NOT chosen here (MVP omits
    # them; idle/sleep/busy/attention is enough to prove the loop).
    if not connected:
        return Pet(state="sleep")
    if prompt is not None and prompt.decidedBy is None:
        return Pet(state="attention")
    if sessions.running >= 3:
        return Pet(state="busy")
    return Pet(state="idle")


def _derive_pet_disconnected() -> Pet:
    return Pet(state="sleep")


def _event_at(event: Event) -> int | None:
    return getattr(event, "at", None)

File: src/src/core/state.ts
/**
 * BuddyState + pure reducer.
 *
 * Imports types directly from schema/state.ts — the single source of truth.
 * The reducer is the only code that transitions state. Everything else reads.
 *
 * Contract (invariants):
 * 1. `version` is monotonic; reduce bumps by exactly 1 on change,
 *    returns the SAME state object otherwise.
 * 2. Reducer is pure: no logging, no time lookup, no I/O.
 *    `at` is always passed in via the event.
 * 3. Prompt replacement is atomic: an event either creates, decides,
 *    or clears the prompt. Decisions persist on prompt.decidedBy
 *    until a RequestCleared event removes them.
 */

import type {
  BuddyState,
  DesktopLink,
  DesktopStatus,
  Pet,
  PendingDecision,
  Prompt,
  SessionCounts,
} from "../../schema/state.js";
import { INITIAL_BUDDY_STATE } from "../../schema/state.js";
import type { Event } from "./events.js";
import type { ActivitySignalKind } from "./events.js";

export { INITIAL_BUDDY_STATE };
export type { BuddyState };

// Per-agent connection tracking (not in schema — internal to daemon).
export interface AgentLink {
  agentId: string;
  status: DesktopStatus;
  lastSeenAt: number | null;
}

// Extended state that includes agent-level tracking.
// The `desktop` field is derived from agents for wire compat.
export interface InternalState extends BuddyState {
  agents: AgentLink[];
  staleMs: number;
  lastActivitySignal: ActivitySignalKind | null;
}

export const INITIAL_STATE: InternalState = {
  ...INITIAL_BUDDY_STATE,
  agents: [],
  staleMs: 30_000,
  lastActivitySignal: null,
};

// ---------------------------------------------------------------------------
// Reducer
// ---------------------------------------------------------------------------

export function reduce(state: InternalState, event: Event): InternalState {
  const next = reduceInner(state, event);
  if (next === state) return state;
  return {
    ...next,
    version: state.version + 1,
    updatedAt: event.at,
  };
}

function reduceInner(state: InternalState, event: Event): InternalState {
  switch (event.type) {
    case "agent_connected":
      return handleAgentConnected(state, event.agentId, event.at);

    case "agent_disconnected":
      return handleAgentDisconnected(state, event.agentId);

    case "request_arrived":
      return handleRequestArrived(state, event);

    case "request_decided":
      return handleRequestDecided(state, event);

    case "request_expired":
    case "request_cleared":
      return handleRequestRemoved(state, event.requestId);

    case "stale_tick":
      return handleStaleTick(state, event.at);

    case "activity_signal":
      return handleActivitySignal(state, event);

    default:
      return state;
  }
}

// ---------------------------------------------------------------------------
// Event handlers
// ---------------------------------------------------------------------------

function handleAgentConnected(
  state: InternalState,
  agentId: string,
  at: number,
): InternalState {
  const existing = state.agents.find((a) => a.agentId === agentId);
  if (existing && existing.status === "connected") return state;

  const agents = existing
    ? state.agents.map((a) =>
        a.agentId === agentId
          ? { ...a, status: "connected" as const, lastSeenAt: at }
          : a,
      )
    : [...state.agents, { agentId, status: "connected" as const, lastSeenAt: at }];

  return applyAgents(state, agents, "agent_connected");
}

function handleAgentDisconnected(
  state: InternalState,
  agentId: string,
): InternalState {
  const existing = state.agents.find((a) => a.agentId === agentId);
  if (!existing || existing.status === "disconnected") return state;

  const agents = state.agents.map((a) =>
    a.agentId === agentId
      ? { ...a, status: "disconnected" as const }
      : a,
  );

  const allDisconnected = !agents.some((a) => a.status === "connected" || a.status === "stale");
  const next = applyAgents(state, agents, "agent_disconnected");
  if (allDisconnected) {
    return { ...next, lastActivitySignal: null };
  }
  return next;
}

function handleRequestArrived(
  state: InternalState,
  event: Extract<Event, { type: "request_arrived" }>,
): InternalState {
  // Update agent liveness.
  const agents = touchAgent(state.agents, event.agentId, event.at);

  const prompt: Prompt = {
    id: event.requestId,
    tool: event.tool,
    hint: event.hint,
    arrivedAt: event.at,
    decidedBy: null,
  };

  const desktop = deriveDesktop(agents);
  const active = agents.some((a) => a.status === "connected");
  const sessions: SessionCounts = { total: 1, running: 1, waiting: 1 };
  const pet = derivePet(prompt, sessions, active, state.lastActivitySignal, null);
  const msg = shortMsg(event.tool, event.hint, event.source);

  return {
    ...state,
    agents,
    desktop,
    sessions,
    prompt,
    pet,
    msg,
    entries: [msg, ...state.entries].slice(0, 10),
    lastSignal: `request_arrived:${event.tool}`,
  };
}

function handleRequestDecided(
  state: InternalState,
  event: Extract<Event, { type: "request_decided" }>,
): InternalState {
  if (!state.prompt || state.prompt.id !== event.requestId) return state;
  if (state.prompt.decidedBy !== null) return state; // first decision wins

  const wireDecision = event.decision === "approve" ? "once" : "deny";
  const decidedBy: PendingDecision = {
    clientId: event.clientId,
    decision: wireDecision as "once" | "deny",
    at: event.at,
  };

  return {
    ...state,
    prompt: { ...state.prompt, decidedBy },
    lastSignal: `request_decided:${event.decision}`,
  };
}

function handleRequestRemoved(
  state: InternalState,
  requestId: string,
): InternalState {
  if (!state.prompt || state.prompt.id !== requestId) return state;

  const sessions: SessionCounts = { total: 0, running: 0, waiting: 0 };
  const active = state.agents.some((a) => a.status === "connected");
  const pet = derivePet(null, sessions, active, state.lastActivitySignal, null);

  return {
    ...state,
    prompt: null,
    sessions,
    pet,
    lastSignal: "request_cleared",
  };
}

function handleStaleTick(state: InternalState, now: number): InternalState {
  let changed = false;
  const agents = state.agents
    .filter((a) => {
      if (a.status === "disconnected" && a.lastSeenAt !== null && now - a.lastSeenAt > state.staleMs) {
        changed = true;
        return false;
      }
      return true;
    })
    .map((a) => {
      if (
        a.status === "connected" &&
        a.lastSeenAt !== null &&
        now - a.lastSeenAt > state.staleMs
      ) {
        changed = true;
        return { ...a, status: "stale" as const };
      }
      return a;
    });

  // Check one-shot expiry.
  const oneShotExpired =
    state.pet.oneShotUntil !== null && now >= state.pet.oneShotUntil;
  if (oneShotExpired) changed = true;

  if (!changed) return state;

  const desktop = deriveDesktop(agents);
  const active = agents.some((a) => a.status === "connected");
  const pet = derivePet(
    state.prompt, state.sessions, active,
    state.lastActivitySignal,
    oneShotExpired ? null : state.pet.oneShotUntil,
  );
  return { ...state, agents, desktop, pet };
}

function handleActivitySignal(
  state: InternalState,
  event: Extract<Event, { type: "activity_signal" }>,
): InternalState {
  const agents = touchAgent(state.agents, event.agentId, event.at);
  const desktop = deriveDesktop(agents);
  const active = agents.some((a) => a.status === "connected");

  const oneShotUntil = event.signal === "celebrate" ? event.at + 2000 : null;

  const pet = derivePet(
    state.prompt, state.sessions, active,
    event.signal, oneShotUntil,
  );

  const msg = `[${event.agentId}] ${event.signal}`;

  return {
    ...state,
    agents,
    desktop,
    pet,
    lastActivitySignal: event.signal,
    lastSignal: event.signal,
    msg,
    entries: [msg, ...state.entries].slice(0, 10),
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function applyAgents(state: InternalState, agents: AgentLink[], lastSignal?: string): InternalState {
  const desktop = deriveDesktop(agents);
  const active = agents.some((a) => a.status === "connected");
  const pet = derivePet(
    state.prompt, state.sessions, active,
    state.lastActivitySignal, state.pet.oneShotUntil,
  );
  const next: InternalState = { ...state, agents, desktop, pet };
  if (lastSignal !== undefined) next.lastSignal = lastSignal;
  return next;
}

function deriveDesktop(agents: AgentLink[]): DesktopLink {
  const connected = agents.filter((a) => a.status === "connected");
  const stale = agents.filter((a) => a.status === "stale");

  let status: DesktopStatus;
  if (connected.length > 0) {
    status = "connected";
  } else if (stale.length > 0) {
    status = "stale";
  } else {
    status = "disconnected";
  }

  // Latest heartbeat across all agents.
  const lastHeartbeatAt = agents.reduce<number | null>((best, a) => {
    if (a.lastSeenAt === null) return best;
    return best === null ? a.lastSeenAt : Math.max(best, a.lastSeenAt);
  }, null);

  return { status, lastHeartbeatAt, secure: false };
}

function derivePet(
  prompt: Prompt | null,
  sessions: SessionCounts,
  active: boolean,
  lastSignal: ActivitySignalKind | null,
  oneShotUntil: number | null,
): Pet {
  if (!active) return { state: "sleep", oneShotUntil: null, species: "bufo" };

  if (prompt !== null && prompt.decidedBy === null)
    return { state: "attention", oneShotUntil: null, species: "bufo" };

  if (oneShotUntil !== null)
    return { state: "celebrate", oneShotUntil, species: "bufo" };

  if (lastSignal === "start_working" || lastSignal === "keep_working")
    return { state: "busy", oneShotUntil: null, species: "bufo" };

  if (lastSignal === "stop_working")
    return { state: "idle", oneShotUntil: null, species: "bufo" };

  if (sessions.running >= 3)
    return { state: "busy", oneShotUntil: null, species: "bufo" };

  return { state: "idle", oneShotUntil: null, species: "bufo" };
}

function touchAgent(agents: AgentLink[], agentId: string, at: number): AgentLink[] {
  const existing = agents.find((a) => a.agentId === agentId);
  if (existing) {
    return agents.map((a) =>
      a.agentId === agentId
        ? { ...a, status: "connected" as const, lastSeenAt: at }
        : a,
    );
  }
  return [...agents, { agentId, status: "connected" as const, lastSeenAt: at }];
}

function shortMsg(tool: string, hint: string, source: string): string {
  const base = source && source !== "other" ? `[${source}] ${tool}` : tool;
  if (!hint) return base;
  const short = hint.length <= 60 ? hint : hint.slice(0, 57) + "...";
  return `${base}: ${short}`;
}

